Задача: Ловим события.

piligrim-curious opened the pull request:
**SpellEvent**: Класс сущности заклинаний;
**HogwartsSpells**: Система управления заклинаниями.

piligrim-curious committed:
Задача: Ловим события.
SpellEvent: Класс сущности заклинаний;
HogwartsSpells: Система управления заклинаниями.

SnurnicynKonstantin commented on src/main/java/school/faang/catching_events/SpellEvent.java:
```diff
@@ -0,0 +1,25 @@
+@Getter
+@EqualsAndHashCode
+public class SpellEvent {
+    private static int nextId = 1;
```
Заведи лучше переменную nextId в HogwartsSpells. Пусть там будет этот счетчик. Когда там через конструктор создаешь `new SpellEvent()` то там и делай:
```
int currentId = currentId++ 
new SpellEvent(currentId, ...)
```
тогда эта модель будет чистинькая, без лишних полей типа nextId

SnurnicynKonstantin commented on src/main/java/school/faang/catching_events/HogwartsSpells.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    static List<SpellEvent> getSpellEventsByType(String eventType) {
+        return new ArrayList<>(spellsByType.getOrDefault(eventType, Collections.emptyList()));
```
Классно, что такой короткой конструкцией возвращаешь Collections.emptyList(), если данных нет.
Можно ли еще короче описать? Например так `return spellsByType.getOrDefault(eventType, Collections.emptyList())`

SnurnicynKonstantin commented on src/main/java/school/faang/catching_events/HogwartsSpells.java:
```diff
@@ -0,0 +1,102 @@
+    private static final Map<Integer, SpellEvent> spellById = new HashMap<>();
+    private static final Map<String, List<SpellEvent>> spellsByType = new HashMap<>();
+
+    static void addSpellEvent(String eventType, String actionDescription) {
```
А почему методы статичными сделал?

piligrim-curious replied:
В данном случае не требовалось создать несколько экземпляров класса, поэтому для простоты я решил сделать все методы статичными. Но я с нейронкой побеседовал и понял, что лучше делать не статичные методы. (Позволяет избежать антипаттерны, проще тестировать, инкапсуляция состояния, гибкость в создании независимых коллекций)

SnurnicynKonstantin commented on src/main/java/school/faang/catching_events/HogwartsSpells.java:
```diff
@@ -0,0 +1,102 @@
+        if (event == null) {
+            throw new IllegalArgumentException("Spell with ID " + id + " not found");
+        }
+        spellById.remove(id);
```
Операция remove должна вернуть результат. Либо удаленное значение вернется, либо null.
Так что можешь в проверке это использовать. Примерно:
```
        SpellEvent spellEvent = spellById.remove(id);
        if (spellEvent != null) {
           ...
        }
```

piligrim-curious committed:
Задача: Ловим события.
SpellEvent: Класс сущности заклинаний;
HogwartsSpells: Система управления заклинаниями.

Изменения:
(SpellEvent.java)
* Удалено статическое поле nextId;
* Удалён метод getNextId();
* Изменен конструктор.

(HogwartsSpells.java)
* Удалены все модификаторы static;
* Добавлено поле nextId;
* Изменён метод addSpellEvent();
* Упрощён метод getSpellEventsByType();
* Изменён метод deleteSpellEvent();
* Теперь в методе main создаётся экземпляр класса HogwartsSpells (app).

SnurnicynKonstantin reviewed (approved):
Вот теперь с внешним счетчиком и чистой моделью аккуратно стало. Молодец)
